streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -235,8 +297,84 @@ private[streaming] class ReceivedBlockTracker(
     }
   }
 
+  /**
+   * Creates a WAL Writer in a separate thread to enable batching of log events.
+   * Exposed for tests.
+   */
+  protected def createBatchWriteAheadLogWriter(): Option[BatchLogWriter] = {
+    if (!WriteAheadLogUtils.isBatchingEnabled(conf)) return None
+    val writer = checkpointDirOption.map(_ => new BatchLogWriter)
+    writer.foreach { runnable =>
+      val thread = new Thread(runnable, "Batch WAL Writer")
+      thread.setDaemon(true)
+      thread.start()
+    }
+    writer
+  }
+
   /** Check if the write ahead log is enabled. This is only used for testing purposes. */
   private[streaming] def isWriteAheadLogEnabled: Boolean = writeAheadLogOption.nonEmpty
+
+  /** A helper class that writes LogEvents in a separate thread to allow for batching. */
+  private[streaming] class BatchLogWriter extends Runnable {
+
+    var active: Boolean = true
+
+    private def writeRecords(records: Seq[ReceivedBlockTrackerLogEvent]): Unit = {
```

Comment:
add scala docs, saying what this does. In fact, its not obvious from the name, what is the functional difference between flushRecords, and writeRecords?